graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -1637,37 +1664,99 @@ def _create_split_files_from_rates(
+                label_column,
+                split_list,
+            )
 
-        # We use multinomial sampling to create a one-hot
-        # vector indicating train/test/val membership
-        def multinomial_sample(label_col: str) -> Sequence[int]:
-            if label_col in {"", "None", "NaN", None}:
-                return [0, 0, 0]
-            return rng.multinomial(1, split_list).tolist()
+            rng = default_rng(seed=seed)
+
+            # We use multinomial sampling to create a one-hot
+            # vector indicating train/test/val membership
+            def multinomial_sample(label_col: str) -> Sequence[int]:
+                if label_col in {"", "None", "NaN", None}:
+                    return [0, 0, 0]
+                return rng.multinomial(1, split_list).tolist()
+
+            group_col_name = "sample_boolean_mask"  # TODO: Ensure uniqueness of column?
+
+            # TODO: Use PandasUDF and check if it is faster than UDF
+            split_group = F.udf(multinomial_sample, ArrayType(IntegerType()))
+            # Convert label col to string and apply UDF
+            # to create one-hot vector indicating train/test/val membership
+            input_col = F.col(label_column).astype("string") if label_column else F.lit("dummy")
+            int_group_df = input_df.select(split_group(input_col).alias(group_col_name))
+            train_mask_df = int_group_df.select(F.col(group_col_name)[0].alias("train_mask"))
+            val_mask_df = int_group_df.select(F.col(group_col_name)[1].alias("val_mask"))
+            test_mask_df = int_group_df.select(F.col(group_col_name)[2].alias("test_mask"))
+        else:
+            # custom node/edge label
+            # create custom mask dataframe for one of the types: train, val, test
+            def process_custom_mask_df(input_df, split_file, mask_type):
+                if mask_type == "train":
+                    file_path = split_file.train
+                elif mask_type == "val":
+                    file_path = split_file.valid
+                elif mask_type == "test":
+                    file_path = split_file.test
+                else:
+                    raise ValueError("Unknown mask type")
+
+                if len(split_file.column) == 1:
+                    custom_mask_df = self.spark.read.parquet(
+                        self.input_prefix + "/" + file_path
+                    ).select(col(custom_split_file.column[0]).alias(f"custom_{mask_type}_mask"))
+                    mask_df = input_df.join(
+                        custom_mask_df,
+                        input_df[NODE_MAPPING_STR] == custom_mask_df[f"custom_{mask_type}_mask"],
+                        "left_outer",
+                    )
+                    mask_df = mask_df.withColumn(
+                        f"{mask_type}_mask",
+                        when(mask_df[f"custom_{mask_type}_mask"].isNotNull(), 1).otherwise(0),
+                    ).select(f"{mask_type}_mask")
```

Comment:
Avoid calling `withColumn` if possible (as this will be called in a for loop), see the note about performance in https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.withColumn.html#pyspark.sql.DataFrame.withColumn

Let's try to re-write this with a `select` statement.